ui/plugins/auto_classification/controller.js

```diff
@@ -1,22 +1,745 @@
+                hasBug: bugNumber ? true : false, // Did the option have a bug when first created
+                matches: matches || null,
+                always: false, // For type = 'ignore' ignore just for this job or
+                               // also for future cases where the same line is matched
+                isBest: false,
+                get icon() {
+                    return this.isBest ? "autoclassified" : 'none';
+                }
+            });
+        };
+        return ThClassificationOption;
+    }
+]);
+
+treeherder.factory('ThStructuredLinePersist', ['$q',
+                                               'thExtendProperties',
+                                               'thValidBugNumber',
+                                               'thNotify',
+                                               'thTabs',
+                                               'ThFailureLinesModel',
+                                               'ThClassifiedFailuresModel',
+    function($q, thExtendProperties, thValidBugNumber, thNotify, thTabs, ThFailureLinesModel,
+             ThClassifiedFailuresModel) {
+        /*
+         When saving a structured line, we need to account for the following cases:
+
+         * An autoclassified failure with an existing bug number is selected.
+           In this case we set the best classification of the failure line to
+           that classification.
+
+         * An autoclassified failure with no existing bug number is selected.
+           In that case we first update the classified failure to set the bug number
+           and then set the best classification of the failure line to that classification
+
+         * An unstructured bug is selected.
+           In that case we first create a new classified failure for the bug (or get the id of
+           an existing one), and then set the best classification to that classification
+
+         * A new bug number is manually entered
+           This works just like selecting a unstructured bug
+
+         * The line is to be ignored now and in the future
+           This works just like selecting an autoclassified failure, with bug number 0.
+           In practice we use the same codepath as for an unstructured bug as this allows
+           providing the bug number (0) rather than the classified failure id as input.
+
+         * The line is ignored this time only
+           In this case we set the best classification of the failure line to null and
+           verify the failure
+
+         */
+
+        var ThStructuredLinePersist = function() {};
+
+        var verifyLine = function(line, cf) {
+            return ThFailureLinesModel.verify(line.id, cf ? cf.id : null)
+                .then(function (response) {
+                    thNotify.send("Classification saved", "success");
+                }, function (errorResp) {
+                    thNotify.send("Error saving classification", "danger");
```

Comment:
you *could* use the extra param to make this sticky.  I often do that on failures to make it more noticeable, but no big deal.  Totally up to you.  :)